wolk/service/firmware_update/FirmwareUpdateService.cpp

```diff
@@ -0,0 +1,292 @@
+                                             DataService& dataService,
+                                             std::shared_ptr<FirmwareParametersListener> firmwareParametersListener,
+                                             FirmwareUpdateProtocol& protocol, const std::string& workingDirectory)
+: m_connectivityService(connectivityService)
+, m_dataService(dataService)
+, m_deviceKey(std::move(deviceKey))
+, m_sessionFile(FileSystemUtils::composePath(SESSION_FILE, workingDirectory))
+, m_firmwareParametersListener(std::move(firmwareParametersListener))
+, m_protocol(protocol)
+{
+}
+
+void FirmwareUpdateService::onBuild()
+{
+    LOG(TRACE) << METHOD_INFO;
+
+    // Check if there is a session going on, and that we have a firmware installer
+    if (FileSystemUtils::isFilePresent(m_sessionFile))
+    {
+        // If a file installer is not present, this means the user had removed a firmware installer, even if we have a
+        // session going on.
+        if (m_firmwareInstaller == nullptr)
+        {
+            LOG(WARN) << "Detected a Firmware Update session but a firmware installer is missing now.";
+            deleteSessionFile();
+            queueStatusMessage(FirmwareUpdateStatus::ERROR, FirmwareUpdateError::UNKNOWN);
+            return;
+        }
+
+        // Read the old version of the firmware
+        auto content = std::string{};
+        if (!FileSystemUtils::readFileContent(m_sessionFile, content))
+        {
+            LOG(WARN) << "Failed to read the content of the session file.";
+            deleteSessionFile();
+            queueStatusMessage(FirmwareUpdateStatus::ERROR, FirmwareUpdateError::UNKNOWN);
+            return;
+        }
+
+        // Ask the installer if the update went well
+        auto success = m_firmwareInstaller->wasFirmwareInstallSuccessful(content);
+        if (success)
+            queueStatusMessage(FirmwareUpdateStatus::SUCCESS);
+        else
+            queueStatusMessage(FirmwareUpdateStatus::ERROR, FirmwareUpdateError::INSTALLATION_FAILED);
+        deleteSessionFile();
+    }
+}
+
+void FirmwareUpdateService::onConnected()
+{
+    LOG(TRACE) << METHOD_INFO;
+
+    // Report the parameters
+    reportParameters();
+
+    // Obtain the parameters if necessary
+    if (m_firmwareParametersListener != nullptr)
+        obtainParametersAndAnnounce();
+
+    // Clean the queue
+    while (!m_queue.empty())
+    {
+        m_connectivityService.publish(m_queue.front());
+        m_queue.pop();
+    }
+}
+
+const Protocol& FirmwareUpdateService::getProtocol()
+{
+    return m_protocol;
+}
+
+void FirmwareUpdateService::messageReceived(std::shared_ptr<Message> message)
+{
+    LOG(TRACE) << METHOD_INFO;
+
+    // Check that the message is not null
+    if (message == nullptr)
+    {
+        LOG(ERROR) << "Failed to process received message -> The message is null.";
+        return;
+    }
+    if (m_firmwareInstaller == nullptr)
```

Comment:
Well, when the device is in `PULL` mode, it would not use this mechanism for Firmware Update. It would just obtain the newest values of parameters regarding the firmware update, and move on. So for that reason, it does not need to receive messages. I thought it was better user experience to say this, but I'll be literal with the error message now. And the comment is again out of line, so my change won't put the comment in `Outdated`.